stdlib/event/gui.lua

```diff
@@ -46,44 +19,144 @@ function Gui.dispatch(event)
 -- @return (<span class="types">@{Gui}</span>)
-function Gui.remove(event_id, gui_element_pattern)
-    fail_if_missing(event_id, "missing event argument")
+function Gui.register(event_ids, gui_element_pattern, handler)
+    fail_if_missing(event_ids, "missing event name argument")
+    fail_if_missing(gui_element_pattern, "missing gui name or pattern argument")
+
+    if type(event_ids) == "table" then
+        for _, v in pairs(event_ids) do
+            Gui.register(v, gui_element_pattern, handler)
+        end
+        return Gui
+    end
 
     if type(gui_element_pattern) ~= "string" then
         error("gui_element_pattern argument must be a string")
     end
 
-    if Event._registry[event_id] then
-        if Event._registry[event_id][gui_element_pattern] then
-            Event._registry[event_id][gui_element_pattern] = nil
+    if handler == nil then
+        return Gui.remove(event_ids, gui_element_pattern)
+    end
+
+    if not Gui._registry[event_ids] then
+        Gui._registry[event_ids] = {}
+        Event.register(event_ids, Gui.dispatch)
+    end
+    if not Gui._registry[event_ids][gui_element_pattern] then
+        Gui._registry[event_ids][gui_element_pattern] = {}
+    else
+        local _, reg_index = table.find(Gui._registry[event_ids][gui_element_pattern], function(v) return v == handler end)
+        local warnmsg = "Gui: handler re-registered for event " .. event_ids .. "[" ..  gui_element_pattern .. "]; "
+        if reg_index and reg_index ~= #Gui._registry[event_ids][gui_element_pattern] then
+            table.remove(Gui._registry[event_ids][gui_element_pattern], reg_index)
+            log(warnmsg ..  "reordering to bottom.")
+            table.insert(Gui._registry[event_ids][gui_element_pattern], handler)
+            Event.register(event_ids, Gui.dispatch)
+            return Gui
+        elseif reg_index then
+            log(warnmsg .. "re-registering event.")
+            Event.register(event_ids, Gui.dispatch)
+            return Gui
+        end
+    end
+
+    table.insert(Gui._registry[event_ids][gui_element_pattern], handler)
+    return Gui
+end
+
+--- Removes the handler with matching gui element pattern from the event.
+-- @tparam defines.events event_ids valid values are `defines.events.on_gui_*` from @{defines.events}, or
+-- a list of them
+-- @tparam string gui_element_pattern (optional) name or string regular expression for a handler to remove
+-- @tparam function handler (optional) handler to deregister from specified Gui events.
+-- @return (<span class="types">@{Gui}</span>)
+function Gui.remove(event_ids, gui_element_pattern, handler)
+    fail_if_missing(event_ids, "missing event argument")
+
+    if type(event_ids) == "table" then
+        for _, v in pairs(event_ids) do
+            Gui.remove(v, gui_element_pattern, handler)
+        end
+        return Gui
+    end
+
+    if gui_element_pattern and type(gui_element_pattern) ~= "string" then
```

Comment:
`if Is.String(gui_element_pattern) then`

Reply:
Fixed downstream... Now several places that incorrectly accept Is.False garbage inputs are starting to look more like bugs and less like "when in rome" situations to me, but that could turn into a pretty big audit so I'm ignoring it for now.